tests/test_restriction.cpp

```diff
@@ -0,0 +1,288 @@
+    gyro::icntl[Param::ntheta_exp]     = (std::get<0>(GetParam()) % 3) + 3;
+    gyro::icntl[Param::DirBC_Interior] = std::get<1>(GetParam());
+    if (gyro::icntl[Param::nr_exp] == 3)
+        gyro::icntl[Param::fac_ani] = 2; //anisotropy should not exceed grid size
+
+    gmgpolar test_p;
+
+    create_grid(test_p); //Mocking create_grid_polar, check_geom and define_coarse_nodes
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int;
+    int cr_int     = test_p.v_level[1]->nr_int;
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta;
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta;
+
+    std::vector<double> u_test(p_level.m);
+    for (int z = 0; z < p_level.m; z++) {
+        u_test[z] = 10 - (2 / (z + 1));
+    }
+
+    std::vector<double> sol = p_level.apply_restriction_bi(u_test);
+
+    /*
+     * The restriction operator accumulates the values in the direct vicinity and stores them in the corresponding node.
+     * We hence calculate 8 values for every coarse node. These 8 values are the fine nodes in the vicinity that are to be accumulated
+     * in the coarse node
+     *
+     * we treat values in r as the x-axis. values in theta as the y-axis. Hence the vector 'adjacent' stores the values in the order:
+     * **bottom_left, left, top_left, bottom, top, bottom_right, right, top_right;**
+     */
+
+    for (int j = 0; j < cr_int + 1; j++) {
+        for (int i = 0; i < ctheta_int; i++) {
+            std::vector<double> adjacent(8, -1.0);
+            if (j == 0) { //interior-most circle- nodes to the left (lower in radii indices) disappear
+                adjacent[0] = 0.0;
+                adjacent[1] = 0.0;
+                adjacent[2] = 0.0;
+            }
+            if (j == cr_int) { //exterior-most circle- nodes to the right disappear
+                adjacent[5] = 0.0;
+                adjacent[6] = 0.0;
+                adjacent[7] = 0.0;
+            }
+
+            int k = 0;
+            std::vector<std::tuple<double, double>> h_p(8, {-1, -1}); // (h_p, h_{p-1}) relative grid sizes
+            std::vector<std::tuple<double, double>> k_q(8, {-1, -1}); // (k_q, k_{q-1})
+            // z and y represent relative positions to the coarse node. e.g. if (z,y)=(-1,1) then we consider the fine node to the top-left
+            for (int z = -1; z < 2; z++) {
+                for (int y = -1; y < 2; y++) {
+                    if (z != 0 || y != 0) {
+                        if (adjacent[k] != 0.0) {
+
+                            adjacent[k] =
+                                u_test[(2 * j + z) * p_level.ntheta_int +
+                                       ((i != 0 || y > -1)
+                                            ? (2 * i + y)
+                                            : p_level.ntheta_int -
+                                                  1)]; //adjacent value. consider periodic boundary in theta direction
```

Comment:
what happens here?

Reply:
We consider all 8 adjacent finde nodes to the coarse node. the fine nodes in r direction have coordinates 2*j-1 and 2*j+1 and in theta direction 2*i-1 and 2*i+1 (if we consider interior nodes). These have to be accumulated in the restriction operator